src/bson-encode.c

```diff
@@ -648,6 +648,44 @@ void php_phongo_zval_to_bson(zval* data, php_phongo_bson_flags_t flags, bson_t*
 	php_phongo_field_path_free(field_path);
 } /* }}} */
 
+/* Converts the argument to a bson_value_t. If the object is an instance of
+ * MongoDB\BSON\Serializable, the return value of bsonSerialize() will be
+ * used. */
+void php_phongo_zval_to_bson_value(zval* data, php_phongo_bson_flags_t flags, bson_value_t* value TSRMLS_DC) /* {{{ */
+{
+	bson_iter_t iter;
+	bson_t      bson = BSON_INITIALIZER;
+
+#if PHP_VERSION_ID >= 70000
+	zval* data_object = ecalloc(sizeof(zval), 1);
+#else
+	zval* data_object = NULL;
+	ALLOC_INIT_ZVAL(data_object);
+#endif
+
+	array_init_size(data_object, 1);
+	add_assoc_zval(data_object, "data", data);
+
+#if PHP_VERSION_ID >= 70000
+	Z_TRY_ADDREF_P(data);
```

Comment:
I noticed that both `encrypt` and `decrypt` tests fail if I switch to `Z_ADDREF_P`. Not calling any `ADDREF_P` macro also causes an error, since we add `data` to the `data_object` zval which is later freed using `zval_ptr_dtor`.

Reply:
Oh, I missed how `data_object` was used in this function. I think this makes sense, given that the value being encrypted in the test file is a string literal. PHP 7 will end up using an [interned string](https://wiki.php.net/rfc/performanceimprovements#interned_strings), which is _not_ ref-counted.

While `keyId` processing also uses this code path, I expect the interned string case never comes up in practice since that value originates from `createDataKey()` and will almost certainly never be a string literal.

This all LGTM.

Reply:
Ah, didn't think about interned strings when using a string literal. Thank you for the explanation! ❤️